### PR TITLE
[mypyc] Add initial support for compiling singledispatch functions

### PR DESCRIPTION
**Pull request by pranavrajpal**
This PR adds initial support for compiling functions marked with singledispatch by generating IR that checks the type of the first argument and calls the correct implementation, falling back to the main singledispatch function if none of the registered implementations have a dispatch type that matches the argument.

Currently, this only supports both one-argument versions of `register` (passing a type as an argument to register or using type annotations), and only works if `register` is used as a decorator.

## Test Plan

Several of the tests in run-singledispatch.test that were previously marked as xfail are now able to pass, so those are changed to regular tests.

### CODE REVIEW

**Review by msullivan — COMMENTED** [review]
This looks really solid, though we should find a way to avoid mutating the AST.

Also, it just occured to me that something like the following test case might be tricky:
```
from functools import singledispatch
from typing import *

@singledispatch
def fun(arg: Any) -> Iterable[int]:
    yield 1

@fun.register(int)
def fun_specialized(arg: str) -> Iterable[int]:
    return [0]
```

Could you add something like that? You can xfail it if it fails and come back to it

File: mypyc/irbuild/function.py [hunk 1]
```diff
@@ -664,13 +691,8 @@ def f(builder: IRBuilder, x: object) -> int: ...
...
 
-    # The environment operates on Vars, so we make some up
-    fake_vars = [(Var(arg.name), arg.type) for arg in rt_args]
-    args = [builder.read(builder.add_local_reg(var, type, is_arg=True), line)
-            for var, type in fake_vars]
-    arg_names = [arg.name if arg.kind in (ARG_NAMED, ARG_NAMED_OPT) else None
-                 for arg in rt_args]
-    arg_kinds = [concrete_arg_kind(arg.kind) for arg in rt_args]
+    arg_info = get_args(builder, rt_args, line)
+    args, arg_kinds, arg_names = arg_info.args, arg_info.arg_kinds, arg_info.arg_names
```
[the remark this answers is not in the document]
Comment (pranavrajpal) [reply mypyc/irbuild/function.py:695]: I mainly did that because I didn't want to accidentally mess up the order of the fields in the assignment.

**Comment by pranavrajpal** [comment]
> Also, it just occured to me that something like the following test case might be tricky

I think having a `register` argument and a type annotation that disagree with each other should really be a mypy error, seeing as it's almost certainly a bug and we probably can't do much in that situation anyway without mypyc raising a `TypeError` about passing in arguments that don't match the type annotation.

Other than that, I think that would be a good test to add. I'll add it to this PR.

**Comment by msullivan** [comment]
Oh, sorry, that part was just a copy/paste screw up. I meant to have no argument to register

**Pull request merged**